src/Core/src/Stream/RewindableStream.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+namespace AsyncAws\Core\Stream;
+
+use AsyncAws\Core\Exception\InvalidArgument;
+
+/**
+ * Provides a Stream that can be read several time.
+ *
+ * @internal
+ *
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+final class RewindableStream implements Stream
+{
+    private $content;
+
+    /**
+     * @var Stream
+     */
+    private $fallback;
+
+    private function __construct(Stream $content)
+    {
+        $this->content = $content;
+    }
+
+    public static function create($content): RewindableStream
+    {
+        if ($content instanceof self) {
+            return $content;
+        }
+        if ($content instanceof Stream) {
+            return new self($content);
+        }
+
+        throw new InvalidArgument(sprintf('Expect content to be a "Stream". "%s" given.', \is_object($content) ? \get_class($content) : \gettype($content)));
+    }
+
+    public function length(): ?int
+    {
+        if (null !== $this->fallback) {
+            return $this->fallback->length();
+        }
+
+        return $this->content->length();
+    }
+
+    public function stringify(): string
+    {
+        if (null !== $this->fallback) {
+            return $this->fallback->stringify();
+        }
+
+        return \implode('', \iterator_to_array($this));
+    }
+
+    public function getIterator(): \Traversable
+    {
+        if (null !== $this->fallback) {
+            yield from $this->fallback;
+
+            return;
+        }
+
+        $resource = \tmpfile();
+        $this->fallback = ResourceStream::create($resource);
+
+        foreach ($this->content as $chunk) {
+            \fwrite($resource, $chunk);
+            yield $chunk;
```

Comment:
If I do : 

```php
$readOnlySteam = // read once stream
echo $readOnlySteam->length(); // 50Mb
$stream = RewindableStream::create($readOnlySteam);
echo $readOnlySteam->length(); // 50Mb

foreach ($stream as $part) {
   break;
}

echo $stream->length(); // 64 kb (or size of chunk)

```

Is this something we should consider?

Reply:
If it wasn't internal, yes, we should check if the stream have been consumed until the end (same for `hash` ans stringify).

Given it's internal, I'm not sure it worth it.

Reply:
Add a comment so we dont forget it in the future. 

> This is for internal use only. If iterating over a stream, the full stream must be consumed. One cannot iterate only over a few items of the stream.

Reply:
done